### PR TITLE
[processing] overlay tools with multiple overlay layers support

### PR DESCRIPTION
**Pull request by alexbruy**
## Description

Add new Intersect, Union and Difference tools which support multiple "overlay" inputs instead of single overlay layer. Tools are working in the following way: pick input and first overlay layer, produce result and use that result as an input at the next step
together with second overlay layer and so on.

For example, given this set of overlapping polygons
image
Difference with multiple overlays will produce this output
image
which is a result of consecutive difference operations between overlays and result of the previous difference

Similarly union will produce single layer
image
 
This is quite useful in models where we can not use loops.

### CODE REVIEW

**Comment by alexbruy** [comment]
Opened as a draft to collect feedback about implementation.

@nyalldawson we have discussed this a bit, maybe you can take a look.

@gioman maybe you can also leave some feedback.

**Comment by gioman** [comment]
@alexbruy many thanks for this! I think that this functionality is very important. I look forward to be able to test it.

**Comment by alexbruy** [comment]
Not stale, awaits feedback.

**Comment by alexbruy** [comment]
@nyalldawson may I ask for your feedback on this?

**Comment by nyalldawson** [comment]
@alexbruy it's on my list to review post-freeze... next week ok?

**Comment by alexbruy** [comment]
@nyalldawson whenever you have time is ok. Thanks!

File: src/analysis/processing/qgsalgorithmmultidifference.cpp [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+        continue;
+
+      QgsVectorLayer *overlayLayer = qobject_cast< QgsVectorLayer * >( layer );
+      if ( !overlayLayer )
+        continue;
+
+      count = 0;
+      if ( i == 0 )
+      {
+        sink.reset( QgsProcessingUtils::createFeatureSink( id, context, sourceA->fields(), geometryType, crs ) );
```
Comment (nyalldawson) [inline src/analysis/processing/qgsalgorithmmultidifference.cpp:145]:
```suggestion
        sink.reset( QgsProcessingUtils::createFeatureSink( QStringLiteral("memory:"), context, sourceA->fields(), geometryType, crs ) );
```

i'd move this down here -- it makes the code slightly more readable (same with line 163)

File: src/analysis/processing/qgsalgorithmmultiintersection.cpp [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+}
+
+QString QgsMultiIntersectionAlgorithm::groupId() const
+{
+  return QStringLiteral( "vectoroverlay" );
+}
+
+QString QgsMultiIntersectionAlgorithm::shortHelpString() const
+{
+  return QObject::tr( "This algorithm extracts the overlapping portions of features in the Input and Overlay layers. "
```
Comment (nyalldawson) [inline src/analysis/processing/qgsalgorithmmultiintersection.cpp:55]:
```suggestion
  return QObject::tr( "This algorithm extracts the overlapping portions of features in the Input and all Overlay layers. "
```

File: src/analysis/processing/qgsalgorithmmultiintersection.cpp [hunk 3]
```diff
@@ -0,0 +1,214 @@
...
+                      "Features in the output Intersection layer are assigned the attributes of the overlapping features "
```
Comment (nyalldawson) [inline src/analysis/processing/qgsalgorithmmultiintersection.cpp:56]:
```suggestion
                      "Features in the output layer are assigned the attributes of the overlapping features "
```

File: src/analysis/processing/qgsalgorithmmultiintersection.cpp [hunk 4]
```diff
@@ -0,0 +1,214 @@
...
+
+QgsProcessingAlgorithm *QgsMultiIntersectionAlgorithm::createInstance() const
+{
+  return new QgsMultiIntersectionAlgorithm();
+}
+
+void QgsMultiIntersectionAlgorithm::initAlgorithm( const QVariantMap & )
+{
+  addParameter( new QgsProcessingParameterFeatureSource( QStringLiteral( "INPUT" ), QObject::tr( "Input layer" ) ) );
+  addParameter( new QgsProcessingParameterMultipleLayers( QStringLiteral( "OVERLAYS" ), QObject::tr( "Overlay layers" ), QgsProcessing::TypeVector ) );
```
Comment (nyalldawson) [inline src/analysis/processing/qgsalgorithmmultiintersection.cpp:68]:
```suggestion
  addParameter( new QgsProcessingParameterMultipleLayers( QStringLiteral( "OVERLAYS" ), QObject::tr( "Overlay layers" ), QgsProcessing::TypeVectorAnyGeometry ) );
```

File: src/analysis/processing/qgsalgorithmmultidifference.cpp [hunk 5]
```diff
@@ -0,0 +1,176 @@
...
+
+QgsProcessingAlgorithm *QgsMultiDifferenceAlgorithm::createInstance() const
+{
+  return new QgsMultiDifferenceAlgorithm();
+}
+
+void QgsMultiDifferenceAlgorithm::initAlgorithm( const QVariantMap & )
+{
+  addParameter( new QgsProcessingParameterFeatureSource( QStringLiteral( "INPUT" ), QObject::tr( "Input layer" ) ) );
+  addParameter( new QgsProcessingParameterMultipleLayers( QStringLiteral( "OVERLAYS" ), QObject::tr( "Overlay layers" ), QgsProcessing::TypeVector ) );
```
Comment (nyalldawson) [inline src/analysis/processing/qgsalgorithmmultidifference.cpp:71]:
```suggestion
  addParameter( new QgsProcessingParameterMultipleLayers( QStringLiteral( "OVERLAYS" ), QObject::tr( "Overlay layers" ), QgsProcessing::TypeVectorAnyGeometry ) );
```

File: src/analysis/processing/qgsalgorithmmultiintersection.cpp [hunk 6]
```diff
@@ -0,0 +1,214 @@
...
+
+    const long total = sourceA->featureCount();
+    QgsOverlayUtils::intersection( *sourceA, *overlayLayer, *sink, context, feedback, count, total, fieldIndicesA, fieldIndicesB );
+  }
+  else
+  {
+    QgsProcessingMultiStepFeedback multiStepFeedback( totalLayerCount, feedback );
+
+    QString id = QStringLiteral( "memory:" );
+    QgsVectorLayer *intersectionLayer;
```
Comment (nyalldawson) [inline src/analysis/processing/qgsalgorithmmultiintersection.cpp:139]:
```suggestion
    QgsVectorLayer *intersectionLayer = nullptr;
```

File: src/analysis/processing/qgsalgorithmmultiintersection.cpp [hunk 7]
```diff
@@ -0,0 +1,214 @@
...
+    outputs.insert( QStringLiteral( "OUTPUT" ), dest );
+
+    const long total = sourceA->featureCount();
+    QgsOverlayUtils::intersection( *sourceA, *overlayLayer, *sink, context, feedback, count, total, fieldIndicesA, fieldIndicesB );
+  }
+  else
+  {
+    QgsProcessingMultiStepFeedback multiStepFeedback( totalLayerCount, feedback );
+
+    QString id = QStringLiteral( "memory:" );
```
Comment (nyalldawson) [inline src/analysis/processing/qgsalgorithmmultiintersection.cpp:138]: same note as previous -- it'd be easier to quickly read the code if this was just hardcoded in lines 167/201

File: src/analysis/processing/qgsalgorithmmultiunion.cpp [hunk 8]
```diff
@@ -0,0 +1,211 @@
...
+  return QStringLiteral( "vectoroverlay" );
+}
+
+QString QgsMultiUnionAlgorithm::shortHelpString() const
+{
+  return QObject::tr( "This algorithm checks overlaps between features within the Input layer and creates separate features for overlapping "
+                      "and non-overlapping parts. The area of overlap will create as many identical overlapping features as there are "
+                      "features that participate in that overlap." )
+         + QStringLiteral( "\n\n" )
+         + QObject::tr( "An Overlay layer can also be used, in which case features from each layer are split at their overlap with features from "
```
Comment (nyalldawson) [inline src/analysis/processing/qgsalgorithmmultiunion.cpp:57]:
```suggestion
         + QObject::tr( "Multiple Overlay layers can also be used, in which case features from each layer are split at their overlap with features from "
```

File: src/analysis/processing/qgsalgorithmmultiunion.cpp [hunk 9]
```diff
@@ -0,0 +1,211 @@
...
+                        "the other one, creating a layer containing all the portions from both Input and Overlay layers. "
```
Comment (nyalldawson) [inline src/analysis/processing/qgsalgorithmmultiunion.cpp:58]:
```suggestion
                        "all other layers, creating a layer containing all the portions from both Input and Overlay layers. "
```

File: src/analysis/processing/qgsalgorithmmultiunion.cpp [hunk 10]
```diff
@@ -0,0 +1,211 @@
...
+
+QgsProcessingAlgorithm *QgsMultiUnionAlgorithm::createInstance() const
+{
+  return new QgsMultiUnionAlgorithm();
+}
+
+void QgsMultiUnionAlgorithm::initAlgorithm( const QVariantMap & )
+{
+  addParameter( new QgsProcessingParameterFeatureSource( QStringLiteral( "INPUT" ), QObject::tr( "Input layer" ) ) );
+  addParameter( new QgsProcessingParameterMultipleLayers( QStringLiteral( "OVERLAYS" ), QObject::tr( "Overlay layers" ), QgsProcessing::TypeVector, QVariant(), true ) );
```
Comment (nyalldawson) [inline src/analysis/processing/qgsalgorithmmultiunion.cpp:71]:
```suggestion
  addParameter( new QgsProcessingParameterMultipleLayers( QStringLiteral( "OVERLAYS" ), QObject::tr( "Overlay layers" ), QgsProcessing::TypeVectorAnyGeometry, QVariant(), true ) );
```

File: src/analysis/processing/qgsalgorithmmultiunion.cpp [hunk 11]
```diff
@@ -0,0 +1,211 @@
...
+
+    QgsOverlayUtils::resolveOverlaps( *sourceA, *sink, feedback );
+    return outputs;
+  }
+  else
+  {
+    QgsProcessingMultiStepFeedback multiStepFeedback( totalLayerCount, feedback );
+
+    QString id = QStringLiteral( "memory:" );
+    QgsVectorLayer *unionLayer;
```
Comment (nyalldawson) [inline src/analysis/processing/qgsalgorithmmultiunion.cpp:129]:
```suggestion
    QgsVectorLayer *unionLayer = nullptr;
```

File: src/analysis/processing/qgsalgorithmmultiunion.cpp [hunk 12]
```diff
@@ -0,0 +1,211 @@
...
+    outputs.insert( QStringLiteral( "OUTPUT" ), dest );
+
+    QgsOverlayUtils::resolveOverlaps( *sourceA, *sink, feedback );
+    return outputs;
+  }
+  else
+  {
+    QgsProcessingMultiStepFeedback multiStepFeedback( totalLayerCount, feedback );
+
+    QString id = QStringLiteral( "memory:" );
```
Comment (nyalldawson) [inline src/analysis/processing/qgsalgorithmmultiunion.cpp:128]: As above

**Comment by nyalldawson** [comment]
@alexbruy 

Sorry for the slow review!

> Opened as a draft to collect feedback about implementation.

It looks good to me :+1:

File: src/analysis/processing/qgsalgorithmmultidifference.cpp (hunk 1 above)
Comment (alexbruy) [reply src/analysis/processing/qgsalgorithmmultidifference.cpp:144]: I see your point, but later in line 147 we have to get layer from the sink id, so it is necessary to keep it in the variable for further use.

File: src/analysis/processing/qgsalgorithmmultiintersection.cpp (hunk 7 above)
Comment (alexbruy) [reply src/analysis/processing/qgsalgorithmmultiintersection.cpp]: Here the same situation, having id in a variable is necessary as later we will use it to retrieve layer.

**Comment by alexbruy** [comment]
@nyalldawson thanks for the review!

**Comment by alexbruy** [comment]
Added tests

**Comment by alexbruy** [comment]
Not stale

**Comment by alexbruy** [comment]
Not stale

**Review by nyalldawson — APPROVED** [review]

**Comment by alexbruy** [comment]
@nyalldawson thanks a lot!

**Comment by gioman** [comment]
Many (many) thanks @alexbruy and @nyalldawson

**Pull request merged**